### ISSUE TITLE
google.com search results are linking doc pages without stylesheets

### ISSUE THREAD

**Issue by hayzey** [angular/angular.js#16432]
**I'm submitting a ...**

- [x] bug report
- [ ] feature request
- [ ] other 

**Current behavior:**

The CSS on the docs page for `angular.element` doesn't seem to be loading: https://docs.angularjs.org/partials/api/ng/function/angular.element.html

![image](https://user-images.githubusercontent.com/12708007/35709010-db046c18-0804-11e8-824c-1b12f26215f3.png)

**Expected / new behavior:**

The page would have CSS applied as usual.

**Minimal reproduction of the problem with instructions:**

Go to https://docs.angularjs.org/partials/api/ng/function/angular.element.html

**AngularJS version:** 1.x.y

stable

**Browser:** [all | Chrome XX | Firefox XX | Edge XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Chrome 63 in macOS

**Anything else:**

These are the only network requests I'm seeing:

![image](https://user-images.githubusercontent.com/12708007/35709102-2a4a7b5a-0805-11e8-970d-df88e49d6a13.png)

**Comment by frederikprijck** [comment]
I don't think u want to use `partials` in the url, but use https://docs.angularjs.org/api/ng/function/angular.element

If you search for `angular.element` on `https://docs.angularjs.org/`, you'll be redirect to `https://docs.angularjs.org/api/ng/function/angular.element` aswell.

I don't think partials are supposed to have css.
If you'd inspect the source, you'd also see this isn't even a valid html page, but it's a partial.

**Comment by Narretz** [comment]
@frederikprijck is right. How did you get the partial url?

**Comment by hayleytom** [comment]
@frederikprijck The CSS loads if I access it with that link.

@Narretz I just searched "angular.element docs" on Google and that was that first link that popped up.

**Comment by frederikprijck** [comment]
@Narretz @hayleytom 

Wow that's correct and that's NOT good.

https://www.google.be/search?q=angular.element+docs&oq=angular.element+docs&aqs=chrome..69i57j69i60l2j69i61j0.198j0j7&sourceid=chrome&ie=UTF-8

**Comment by Narretz** [comment]
I see. Something went wrong with the robots.txt during migration I assume

**Comment by petebacondarwin** [comment]
Weird the robots.txt is:

```
User-agent: *

Disallow: /components/
Disallow: /examples/
Disallow: /img/
Disallow: /js/
Disallow: /partials/
Disallow: /ptore2e/
Disallow: /*.js$
Disallow: /*.map$
Disallow: /Error404.html
```

which looks correct to me. Any ideas?

**Comment by petebacondarwin** [comment]
Perhaps there was a random crawling during the switchover to Firebase Hosting that missed the robots file? Can we trigger a new crawl @Narretz - I don't have access to the Search console for this site.

**Comment by Narretz** [comment]
I just pushed an update that added this robots.txt. 😊 Before there was actually none. I'll look if I have access.

**Comment by Narretz** [comment]
@petebacondarwin I don't have docs.angularjs.org in the search console either. Probably @IgorMinar must trigger the re-crawl.

**Comment by IgorMinar** [comment]
It's strange that the crawler is not honoring the #! contract. We are correctly setting the meta tag.

I don't know if this is related, but I was just saw a notification that  https://code.angularjs.org/snapshot/docs/js/all-versions-data.js is being banned via robots.txt and this prevents the crawler that indexes the site after the js executed from working.

**Comment by IgorMinar** [comment]
some other issues I found:
- the sitemap is missing: https://docs.angularjs.org/sitemap.xml
- robots.txt is missing: https://docs.angularjs.org/robots.txt
- the _escaped_fragment_ url doesn't seem to be returning any content either e.g. https://docs.angularjs.org/?_escaped_fragment_=api%2Fng%2Ffunction%2Fangular.bind just redirects to /api. this seems wrong according to https://developers.google.com/webmasters/ajax-crawling/docs/getting-started. but I'm not sure if I got the url to test with right.

@Narretz can you please take a look?

**Comment by IgorMinar** [comment]
except for robots.txt the other two issues still seem unresolved.

@Narretz I though you said that the sitemap was fixed a few days ago, did I misunderstand you?

**Comment by Narretz** [comment]
@IgorMinar I did, and when I checked it was there. Not sure what's going on. The sitemap is visible on the docs folders for the snapshots. (Not that we need them there, just saying that the build produces them)

**Comment by Narretz** [comment]
Ah I see, it was copied to the deploy folder. Fixing right now ...

**Comment by Narretz** [comment]
@IgorMinar the site map is up: https://docs.angularjs.org/sitemap.xml

The link to the ajax crawling scheme above says this is deprecated. I assume as long as we allow the js crawler access we don't need this.

And it looks like the crawler hasn't reindexed the site at all. Because at least the partial url should not be in the results but it still is.

**Comment by IgorMinar** [comment]
I requested a recrawl using the sitemap via search console.

the issue with partials is that we used to use the fragment scheme before google had a js-enabled crawler, now that js-enabled crawler is a thing the fragment scheme is in the way... :-(

we spent a ton of time fixing js-crawler related issues for angular.io and I'm not sure if we want to go though the same effort on docs.angularjs.org. it would be better if we could just make the fragment scheme work well enough and  not touch it any more.

**Comment by Narretz** [comment]
Okay, I actually wasn't aware that we used the escaped fragment rule, and so there is no rewrite rule (yet) for serving the escaped fragment. Shouldn't take too long to add it. 

However, the site also says the ajax schemes will be discontinued in summer 2018, so I think we still need the site to be crawlable by the JS bot.

**Comment by IgorMinar** [comment]
right, but I think we are now in the state where the escaped fragments are considered as results to be served to users, that's why we see bad search results.

We should restore the escaped fragments functionality so that the results go back to normal - even though this functionality has been deprecated by google's crawler team. supporting the js-crawler is a big undertaking especially compared to restoring the escaped fragment route.

**Comment by Narretz** [comment]
Okay, so here's what we need to do afaict:

- googlebot goes to a docs page like https://docs.angularjs.org/api/ng/function/angular.element 
- it sees that the site has `<meta name="fragment" content="!">`.
- it will request https://docs.angularjs.org/api/ng/function/angular.element?_escaped_fragment_=
- from which we are supposed to serve a full HTML snapshot
- since we don't have a full snapshot, we serve the contents of `/partials/api/ng/function/angular.element`
- we do this with a cloud function, because Firebase doesn't support capture in rewrites (see https://groups.google.com/forum/#!topic/firebase-talk/MW9BlBdVwDY)

Sound good?

**Comment by Narretz** [comment]
Clarification:

- in the function, we have to download the partial file and serve it
- we cannot use a redirect, because the partials url should not get indexed by the crawler
- the download could either be done directly from https://docs.angularjs.org/partials/api/ng/function/angular.element.html or via gcs from the snapshot-stable folder

Not sure what the better approach is.

**Comment by IgorMinar** [comment]
yeah. that sounds good, I think.

I don't remember if google requests: `https://docs.angularjs.org/api/ng/function/angular.element?_escaped_fragment_=` or if the query param has some value. I think you got it right. There was some weirdness about this because of our use of "html5" urls rather than hashbang urls. Can you look at the old server config to confirm that this is right?

with regards to serving - since there is a finite number of urls to serve, wouldn't it be simpler to have dgeni generate the firebase rewrites and then we don't need to deal with serving these files ourselves via functions or what not. Dgeni already does something very very similar when it generates the sitemap. I'm just looking for the most reliable and low maintenance solution...

**Comment by Narretz** [comment]
If there's no hash in the url there is no value in the escaped_fragment
parameter. There is a section in the guide for that:
https://developers.google.com/webmasters/ajax-crawling/docs/getting-started#3-handle-pages-without-hash-fragments

The old setup has "capture" for the parameter value, but like the guide
says, it doesn't seem to be necessary.

You mean dgeni should create a rewrite source / destination pair for every
piece of content we have which we will then insert into the firebase.json?

Unfortunately, Firebase rewrites do not support query parameters.
Everything after ? was ignored in my tests. So we probably need a cloud
function that handles all content sites that checks if the parameter is in
the url. Less then ideal of course.

Igor Minar <notifications@github.com> schrieb am Sa., 10. Feb. 2018 15:19:

**Comment by Narretz** [comment]
Okay, so ? is a glob wildcard for a single character. However, it doesn't seem possible to escape it.

**Comment by IgorMinar** [comment]
OK. So let's go with the cloud function.

Can we get this fixed asap. We are starting to see increased number of people being affected by this. See:  https://news.ycombinator.com/item?id=16353676

**Comment by Narretz** [comment]
PR at https://github.com/angular/angular.js/pull/16451

**Comment by Narretz** [comment]
The pr was merged but there was a problem with the deployment. I'll fix it tomorrow

**Comment by IgorMinar** [comment]
uh. can someone else help to deploy this? @gkalpak ?

**Comment by gkalpak** [comment]
#16452 should fix it.

**Comment by gkalpak** [comment]
Still failed: https://travis-ci.org/angular/angular.js/jobs/340687854#L711

**Comment by Narretz** [comment]
The firebase deployment failed? (Because the job passed). Anyway let's call it a night. I'll take a look tomorrow

**Comment by gkalpak** [comment]
It failed, because the firebase functions dependencies are not installed (and are apparently necessary).
I'm giving it another try: #16453

**Comment by Narretz** [comment]
Thanks to @gkalpak the snapshots are here! https://docs.angularjs.org/guide/databinding?_escaped_fragment_=

**Comment by gkalpak** [comment]
Finally, deployed: https://travis-ci.org/angular/angular.js/builds/340741409#L735 :tada:

**Comment by IgorMinar** [comment]
yay! awesome! thanks @gkalpak 

I tried to confirm that it actually worked, but the search console ui is confusing for the urls crawled via "escaped fragment" method. This is what I get:

![screen shot 2018-02-12 at 4 39 32 pm](https://user-images.githubusercontent.com/216296/36127648-5aff44ae-1013-11e8-8b80-4633b7915382.png)


Note the the UI actually renders fine via the js-enabled-crawler, except that the crawler is not allowed to index that view because several URLs are being black listed in the robots.txt. @Narretz can you please fix them? This is the list:


URL | Type | Reason | Severity |  
-- | -- | -- | -- | --
https://docs.angularjs.org/js/angular-bootstrap/dropdown-toggle.min.js | Script | Blocked | High | robots.txt Tester
https://docs.angularjs.org/js/current-version-data.js | Script | Blocked | High | robots.txt Tester
https://docs.angularjs.org/js/pages-data.js | Script | Blocked | High | robots.txt Tester
https://docs.angularjs.org/js/nav-data.js | Script | Blocked | High | robots.txt Tester
https://docs.angularjs.org/js/docs.min.js | Script | Blocked | High | robots.txt Tester
https://docs.angularjs.org/components/google-code-prettify-1.0.1/src/prettify.js | Script | Blocked | Medium | robots.txt Tester
https://docs.angularjs.org/components/marked-0.3.6/marked.min.js | Script | Blocked | Low | robots.txt Tester
https://docs.angularjs.org/components/lunr-0.7.2/lunr.min.js | Script | Blocked | Low | robots.txt Tester
https://docs.angularjs.org/components/google-code-prettify-1.0.1/src/lang-css.js | Script | Blocked | Low | robots.txt Tester
https://docs.angularjs.org/img/angularjs-for-header-only.svg | Image | Blocked | Low | robots.txt Tester

**Comment by IgorMinar** [comment]
If I'm testing it correctly, then I think the fix in prod is working well. Example:

url: https://docs.angularjs.org/api/ng/directive
_escaped_fragment url: https://docs.angularjs.org/api/ng/directive?_escaped_fragment_=

both work!

**Comment by Narretz** [comment]
> both work!

I would have been devastated if only one of these worked after all this :>

@IgorMinar I've updated the robots.txt to allow access to the js and images that are used by the docs app. (https://github.com/angular/angular.js/commit/96bee0c6591815a41d1fdaced0e25203bd5ac430) I've tested it with this tool: https://technicalseo.com/seo-tools/robots-txt/ That's not google specific though.

Can you also please check what the crawler sees for the direct partials/ urls like https://docs.angularjs.org/partials/api/ng/function/angular.element.html ? Because the current robots.txt excludes them, but the site is still indexed ...

**Comment by IgorMinar** [comment]
Martin, I sent you invite to search console so that you can take a look at
what it looks like there.

**Comment by Narretz** [comment]
The googlebot can now crawl the docs pages better then before, but now it reports the partials/ as blocked :/ I need to find out if it's safe to allow the partials.

<img width="954" alt="angularelementfetch" src="https://user-images.githubusercontent.com/1153097/36200144-035dd3ea-117c-11e8-89d3-77fd60d75bf2.PNG">

(Sorry, the page is in German)

**Comment by IgorMinar** [comment]
The results seem to be improving:

2018-02-13 at 11:28am PT:
![screen shot 2018-02-13 at 11 28 48 am](https://user-images.githubusercontent.com/216296/36214761-292293bc-115f-11e8-8777-8ad9925e3c39.png)

2018-02-14 at 8:14am PT:
<img width="692" alt="screen shot 2018-02-14 at 8 14 08 am" src="https://user-images.githubusercontent.com/216296/36214848-54167804-115f-11e8-9048-4cb5e24b7acb.png">

**Comment by IgorMinar** [comment]
@Narretz I think it's fine to unblock the partial html. Since we provide a sitemap, the crawler should understand that that html is not a url we want to publicize.

**Comment by IgorMinar** [comment]
The traffic is still down, but there seems to be a slight hint of improvement. We need more data to know for sure...

<img width="1378" alt="screen shot 2018-02-14 at 8 21 58 am" src="https://user-images.githubusercontent.com/216296/36215191-478985b2-1160-11e8-8990-a6f145aa7b30.png">

**Comment by Narretz** [comment]
@IgorMinar good to see the traffic is recovering.

For the partials I will add a noindex header which should complement the sitemap

**Comment by IgorMinar** [comment]
sounds good

**Comment by Narretz** [comment]
The original search for angular.element docs now returns a good result:

https://www.google.be/search?q=angular.element+docs&oq=angular.element+docs&aqs=chrome..69i57j69i60l2j69i61j0.198j0j7&sourceid=chrome&ie=UTF-8

unblocking the partials from the crawler has done the trick I think. I have requested a re-index of docs.angularjs.org/api and its direct links.

For some reason the crawler cannot render docs.angularjs.org/ correctly.
And we can also unblock the examples, because the crawler wants to render the iframes as well. This is reported as low priority though.

**Comment by IgorMinar** [comment]
hourly stats:
<img width="1295" alt="screen shot 2018-02-21 at 10 21 36 am" src="https://user-images.githubusercontent.com/216296/36497747-0b880516-16f1-11e8-8a57-8201fc9e42ae.png">

daily stats:
<img width="1276" alt="screen shot 2018-02-21 at 10 28 35 am" src="https://user-images.githubusercontent.com/216296/36498051-04ada858-16f2-11e8-8bb7-7bb1470d76a5.png">

we are slowly recovering but we are not back to normal yet. it looks like we are still ~20% off for daily session count.

**Comment by IgorMinar** [comment]
I don't think that there is any action we need to take. Let's just monitor this further.

**Comment by IgorMinar** [comment]
one more graph that shows that we are recovering this time from search console - note that this one is delayed - the last datapoint is from Feb 18:

<img width="928" alt="screen shot 2018-02-21 at 10 33 53 am" src="https://user-images.githubusercontent.com/216296/36498464-32b68c28-16f3-11e8-8434-0fc8b17795ee.png">

**Comment by Narretz** [comment]
This is resolved - the numbers are still lower than before the migration, but that is possibly an affect of the LTS announcement.
